dominion/src/dominion/arena.rs

```diff
@@ -115,88 +119,103 @@ impl Arena {
+                }
+            }
+            TurnPhase::Buy { .. } => Err(Error::WrongTurnPhase),
+        }
+    }
+
+    pub fn play_treasure(&mut self, player_id: usize, card: CardKind) -> Result<()> {
+        self.check_active(player_id)?;
+
+        let card_index = self.players[player_id]
+            .hand
+            .iter()
+            .position(|&x| x == card)
+            .ok_or_else(|| Error::InvalidCardChoice)?;
+
+        let gains = card.treasure().ok_or_else(|| Error::InvalidCardChoice)?;
+
+        match &mut self.turn.phase {
+            TurnPhase::Action { .. } => Err(Error::WrongTurnPhase),
             TurnPhase::Buy {
                 remaining_buys,
                 total_wealth,
-            } => match location {
-                Location::Hand => match self.players[player_id].hand.remove_item(&card) {
-                    Some(card) => match card.treasure() {
-                        Some(i) => {
-                            *total_wealth += i;
-                            self.players[player_id].play_zone.push(card);
-                            Ok(())
-                        }
-                        None => {
-                            self.players[player_id].hand.push(card);
-                            Err(Error::InvalidCardChoice)
-                        }
-                    },
-                    None => Err(Error::InvalidCardChoice),
-                },
-                Location::Supply => match self.supply.get_mut(card) {
-                    Some(supply_count) => {
-                        if *supply_count == 0 {
-                            Err(Error::NoMoreCards)
-                        } else if *remaining_buys == 0 {
-                            Err(Error::NoMoreBuys)
-                        } else if *total_wealth < card.cost() {
-                            Err(Error::NotEnoughWealth)
-                        } else {
-                            *supply_count -= 1;
-                            *remaining_buys -= 1;
-                            *total_wealth -= card.cost();
-                            self.players[player_id].discard_pile.push(card);
-                            Ok(())
-                        }
-                    }
-                    None => Err(Error::InvalidCardChoice),
-                },
-                _ => Err(Error::InvalidCardLocation),
-            },
+            } => {
+                let player = &mut self.players[player_id];
+                let card = player.hand.remove(card_index);
+                player.play_zone.push(card);
+
+                *total_wealth += gains;
+
+                Ok(())
+            }
+        }
+    }
+
+    pub fn buy_card(&mut self, player_id: usize, card: CardKind) -> Result<()> {
+        self.check_active(player_id)?;
+
+        let mut supply_count = self
+            .supply
+            .get_mut(card)
+            .ok_or_else(|| Error::InvalidCardChoice)?;
+
+        match &mut self.turn.phase {
```

Comment:
This turn phase thing happens often, you can abstract it
```rust
impl TurnPhase {
fn as_buy_phase_mut(&mut self) -> Result<(&mut remaining_buys, &mut total_wealth), Error::WrongTurnPhase> {
match self {
    TurnPhase::Buy { remaining_buys, total_wealth } => Ok((remaining_buys, total_wealth)),
    _ => Err(Error::WrongTurnPhase),
}
}

Instead of the tuple you can have a struct `BuyPhase` with the fields and change the variants to `Buy(BuyPhase)` etc.
```